### DISCUSSION TITLE
WebVTT and Typing issues with Javascript SDK

### DISCUSSION THREAD

**Discussion by nickottocode** [deepgram/community#616] [General help] [answered]
I am writing to get some help with an error I am encountering with the vtt generation code, as well as some small requests to help with developing in the new v3 experience. 

While using the Javascript SDK to interact with your speech to text services and create a WebVTT file from the deepgram response object, I keep encountering an error that there is an undefined property on utterances which cannot be read (I do have utterances turned on / set to true - see provided response output). I think it may have something to do with the v2 -> v3 change going on?

Also, your most recent documentation suggests that developers destructure the response object into { result, error } instead of capturing the deepgramResponseObject as one unit. (link to documentation where it shows the difference in suggested implementations: https://developers.deepgram.com/docs/js-sdk-v2-to-v3-migration-guide#url-file-transcription)

In principle, I am fine with that and it makes sense, except for that the only type I can import from your sdk here is the top level SyncPrerecordedResponse and not the types underlying it like 'Result', 'Alternative', 'Channel', 'Utterance' - things that would be helpful to make it work in the suggested method with destructuring.

It is also not possible to call the other functions exposed through the sdk as functions like `webvtt()` using this method as those functions only accept the higher level type of DeepgramResponse<T>. (technically the `webvtt` function takes `any` but the converter that it instantiates in the call to `parseInput` assumes that `any` value fits the `DeepgramResponse` shape)

To recap (and sorry for having so many things in one message):
- There appears to be an error with the webvtt function throwing on type issues.
- The parameter typing of the functions from the sdk doesn't align with the types from the proposed destructuring of the response object.
- It would be awesome if you could export more types so that when we do destructure in keeping with your v3 documentation we have access to the type definitions of those elements (and their nested / subtypes). That way those response parameters can be captured in the destructuring pattern and passed into functions requiring specific types.

Please let me know if there are any questions or other information needed.

Thank you!

For the first issue, here is some extra info:

Sample code excerpt:

```js
const deepgram = createClient(env.DEEPGRAM_API_KEY)
const result = await deepgram.listen.prerecorded.transcribeUrl(
  { url: 'my-signed-url' },
  { model: 'nova-2', punctuate: true, utterances: true, smart_format: true },
);
const vttContents = webvtt(result); // The error comes from inside the webvtt function
```

Resultant error:

```
{
  "message": "Cannot read properties of undefined (reading 'utterances')",
  "stack": "TypeError: Cannot read properties of undefined (reading 'utterances')
      at DeepgramConverter.getLines (/...my-path/node_modules/@deepgram/captions/src/converters/DeepgramConverter.ts:12:17)
      at webvtt (/...my-path/node_modules/@deepgram/captions/src/index.ts:27:22)
      at DeepgramTranscribeProvider.generateAndUploadCaptions (/...my-path/src/deepgram.ts:90:33)
      at processTicksAndRejections (node:internal/process/task_queues:96:5)
      at async MediaTranscriptionWorkflow.execute (/...my-path/src/workflow.ts:26:31)
      at async MediaTranscriptionSQSController.handleMessage (/...my-path/src/controller.ts:36:5)",
  "name": "TypeError"
}
```

This is not the only file I've tried, just one I can share. This behavior is happening for all attempts to use the webvtt function across multiple files.  Example file and specific results shared over email.

Versions and stuff:
 - Deepgram SDK version: 3.1.9
 - Node version: 20.10.0
 - Typescript version: 5.3.3

CC @samuel-greenberg

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by davidvonthenen** [comment]
I am not super super familiar with the JS SDK and @lukeocodes will probably have more detail. Are you familiar with this project https://github.com/deepgram/deepgram-js-captions ? There was effort to separate the SRT and VTT creation into a separate project in order to decouple and not places dependencies on each other. So in a nutshell you generate a transcription with the utterance on, you can then feed the raw JSON into the JS caption library. You shouldn't need to pass SDK specific classes into the caption library. (At least this is how the Python caption library works). For library specific things, I think Luke is probably a better contact for those questions.

**Comment by lukeocodes** [comment, marked as answer]
Thanks @dvonthenen, the `deepgram-js-captions` is actually a dependency of the SDK, and they're technically using it by utilising those functions of the SDK. If unused, it gets tree-shaken from the modules at build.

I'll address the points in order.

> There appears to be an error with the webvtt function throwing on type issues.

I believe the issue here is that the object is not destructed as prescribed in our docs. Try this @samuel-greenberg

```js
const deepgram = createClient(env.DEEPGRAM_API_KEY)
const { result, error } = await deepgram.listen.prerecorded.transcribeUrl(
  { url: 'my-signed-url' },
  { model: 'nova-2', punctuate: true, utterances: true, smart_format: true },
);
const vttContents = webvtt(result); // The error comes from inside the webvtt function
```

> The parameter typing of the functions from the sdk doesn't align with the types from the proposed destructuring of the response object.

I'm going to need some more info on this one. I am struggling to understand what doesn't align. This would cause no end of issues, and we have a fairly sizable chunk of this logic running tests, and it used extensively in the wild.

> It would be awesome if you could export more types so that when we do destructure in keeping with your v3 documentation we have access to the type definitions of those elements (and their nested / subtypes). That way those response parameters can be captured in the destructuring pattern and passed into functions requiring specific types.

All types are exported from the top file. For e.g.

```ts
import type { DeepgramResponse } from "@deepgram/sdk";

// or

import { createClient, DeepgramResponse } from "@deepgram/sdk";

// or

import { createClient, LiveTranscriptionEvents } from "@deepgram/sdk";
```

See how we round up all helpful exports [here](https://github.com/deepgram/deepgram-js-sdk/blob/main/src/index.ts#L23-L31), which in turn point at round up files like the types index file [here](https://github.com/deepgram/deepgram-js-sdk/blob/main/src/lib/types/index.ts).
